### PR TITLE
fix for RavenDB-12784

### PR DESCRIPTION
**Pull request by myarichuk**
During graph queries that need to "wait for non-stale", make sure to generate auto-indexes.
In order to do that, DynamicQueryRunnier::MatchIndex() is used to get relevant index to wait for "non-stale", and generate auto-index if needed.
Also added precaution so if the query runner will attempt to close transaction for a query that is a part of QueryQueryStep it will properly throw.

### CODE REVIEW

File: src/Sparrow/Json/BlittableJsonReaderObject.cs [hunk 1]
```diff
@@ -26,6 +26,10 @@ public unsafe class BlittableJsonReaderObject : BlittableJsonReaderBase, IDispos
         private readonly bool _isRoot;
         private byte* _objStart;
 
+        #if DEBUG || VALIDATE
```
Comment (talweiss1982) [inline src/Sparrow/Json/BlittableJsonReaderObject.cs]: Is this debug leftovers?

Comment (myarichuk) [reply src/Sparrow/Json/BlittableJsonReaderObject.cs]: yes, but I'd like to leave this - as it may be useful for debugging memory issues in the future

Comment (ayende) [reply src/Sparrow/Json/BlittableJsonReaderObject.cs]: No, lte's remove that, debug builds are already expensive enough

**Review by ayende — CHANGES_REQUESTED** [review]
We need a test case

**Review by ppekrol — APPROVED** [review]

**Comment by ppekrol** [comment]
@myarichuk can you check failing tests?

**Comment by myarichuk** [comment]
The most recent commit should handle the failing tests.

**Pull request merged**